Use WordPress Coding Standards and Scripts

gregrickaby opened the pull request:
Closes #614

- [x] Copy over all WordPress config files
- [x] Remove WDS config files
- [x] Refactor Webpack config(s)
- [x] Copy over NPM Scripts and pass in `--config` flags
- [x] Test dev mode
- [x] Verify source-maps work in dev mode
- [x] Verify Prettier when writing SCSS and JavaScript
- [x] Verify Stylelint when writing SCSS
- [x] Verify LiveReload
- [x] Verify PHPCS via Composer
- [x] Test production/build mode
- [x] Verify asset minification
- [x] Theme wide formatting changes with new standards
- [x] Update README

### DESCRIPTION

This PR introduces a nearly 1:1 sync with [WordPress](https://github.com/WordPress/gutenberg) standards and tooling. The goal is to reduce internal maintenance of standards, documentation, and tooling and just leverage what WordPress already has in place.

- Remove WDS Coding Standards for JavaScript, SCSS, and CSS (PHP standards are still WDS)
- Remove WDS Scripts for [@wordpress/scripts](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-scripts/#setup) defaults
- Remove extra devDependencies by leveraging what already comes with @wordpress/scripts (Autoprefixer, PostCSS, etc...)
- Remove CSS Nano in PostCSS config in favor of `css-minimizer-webpack-plugin` (which uses CSS Nano)
- Add support for (WordPress's) Prettier
- Add support for markdown linting
- Add support for package.json linting
- Add support for better globbing and NPM script matching with Lefthook
- Add support for development vs. production modes (via separate webpack configs and ENV vars) to decrease dev build times (e.g., don't minify stuff in dev mode)

### SCREENSHOTS

Initial _dev_ build times using `npm run dev` are ~2.5 seconds 👍🏻 

![screenshot](https://dl.dropbox.com/s/pf4s386o2bavl6x/Screen%20Shot%202021-03-30%20at%2012.24.42.png?dl=0)

### STEPS TO VERIFY

**Note: Buddy assertions will fail until after this PR is merged and Buddy is updated!**

#### Install
- Check out this PR
- Run `rm -rf node_modules package-lock.json`
- Run `npm i --legacy-peer-deps`

#### Develop

- Run `npm run dev` and make changes SCSS files and monitor build times.
- Run `npm run build` to create a production build and verify everything has been minified/optimized

### POST-MERGE CHORES

- [x] Update Buddy pipelines
- [x] Update WDS Guide coding standards to just point at [WordPress](https://codex.wordpress.org/WordPress_Coding_Standards)
- [x] Remove [WDS packages from NPM repo](https://www.npmjs.com/~webdevstudios) 
- [x] Update Audit docs
- [x] Chat with @aubreypwd about ongoing maintenance for PHP standards (I know we want short array syntax, does this need to be a separate "package" or can it just be added to this repo's `.phpcs.xml.dist` instead? For context, I had to slightly modify the `.eslintrc.js` [file](https://github.com/WebDevStudios/wd_s/pull/618/commits/a2b96c5556ebccaa88a54da513387d520069671a) because we don't use React. I think this is fine that it's in this repo, and doesn't require WDS to create a seperate NPM package.

gregrickaby committed:
use scripts from project template

gregrickaby committed:
check in all wordpress-based config files

gregrickaby committed:
remove rulesets

gregrickaby committed:
remove future declarations

gregrickaby committed:
remove cssnano. leverage preset-env instead

gregrickaby committed:
refactor webpack config

gregrickaby committed:
create production webpack config

gregrickaby committed:
refactor package.json

gregrickaby committed:
install dependencies

gregrickaby committed:
drop sass-lint support

gregrickaby committed:
make it clear the configs are for different modes

gregrickaby committed:
remove autoprefixer and postcss dev dependencies

gregrickaby committed:
remove unused folders

gregrickaby committed:
add wds specific options

gregrickaby committed:
eslint and prettier formatting

gregrickaby committed:
update path

gregrickaby committed:
add markdownlint to recommended extensions

gregrickaby committed:
move license and contributing to root (per github)

gregrickaby committed:
fix link. formatting.

gregrickaby committed:
bump version number

coreymcollins commented:
Thanks for this @gregrickaby!

It looks like this removes Browsersync – are we pulling that out again?

gregrickaby commented:
@coreymcollins 😬  That was an oversight on my part. I'll get it added back in. Sorry!

gregrickaby committed:
add browser-sync support

coreymcollins commented:
In doing some light testing, I've found some things that aren't happening right now:
- Language files are never created in the `/build/` directory, even during `npm run build`
- The `sprite.svg` file isn't being created during `npm run dev`
    - If I change an SVG file while running this, the SVG file IS updated in build but the sprite is never built
- Browsersync starts, but we're not actually running `start` so nothing is being listened to/rebuilt

I think there are a couple of additional packages we'll need to install, if we want to keep the current workflow:
- `npm-run-all`
- `onchange`

`npm-run-alll` lets us use `run-p` (and `run-s`, which may be useful) so we can setup a `start:sync` type of script like we have running currently (`"start:sync": "run-p start sync watch:icons",`).

`onchange` helps us listen for changes in the `src/image/icons` directory so we can regenerate the `sprite.svg` file when something changes. Without the sprite, nothing will show up when using the SVG display PHP functions during dev.

I'm sure there's more than one way to do these things, so I wanted to note them all here so we can chat about the best route to go with these rather than hacking at your branch and making changes.

gregrickaby committed:
add pot command

gregrickaby committed:
add svg spritemap plugin. remove mode.

gregrickaby committed:
add gifsicle support. remove mode.

gregrickaby committed:
install dependencies

gregrickaby committed:
bump

gregrickaby committed:
refactor scripts. drop entry. add cross-env & svg-spritemap

Commit:
Merge branch 'main' into feature/coding-standards

gregrickaby commented:
@coreymcollins I've taken care of the issues you mentioned.

1. I deferred the SVG sprite process to Webpack. We no longer need an NPM script or the `onchange` package. Webpack will create the sprite when running either `npm run start` or `npm run build`.
2. I added back the `composer run-script pot` to generate the `_s.pot` file.
3. I added `npm-run-all` and refactored a few NPM scripts to take advantage of `run-s` or `run-p`.
4. I'm a bit nostalgic, so I brought back `npm run watch` which will kick-off Live Reload and Browsersync. If an engineer doesn't want to use Browsersync, then run `npm run start` instead.
5. After studying WordPress's [webpack.config.js](https://github.com/WordPress/gutenberg/blob/trunk/packages/scripts/config/webpack.config.js#L27), I noticed they have optimizations set up based on ENV_VARS. As such, I added the `cross-env` package and [passed ENV=* ](https://github.com/WebDevStudios/wd_s/blob/feature/coding-standards/package.json#L43) to both `npm run start` and `npm run build`. Now we can take advantage of those native optimizations. An alternative to using the `cross-env` package, would be to create a `.env` file.
6. I removed `enabled: true` from `tailwind.config.js`. Because we're passing `ENV=production` now, Tailwind [automatically purges for production](https://tailwindcss.com/docs/optimizing-for-production#basic-usage) and not during development. I was able to work in a PHP template and see my Tailwind changes in ~1 second using Browsersync! This fixes @oliverharrison issue where he had to restart build scripts each time he added a Tailwind class to PHP template. 
7. Updates README and addresses #602 

During my testing tonight, Live Reload and Browsersync are working great, and sprite and image optimizations look correct. 

- Live Reload times are still around ~2.5 seconds
- Browsersync times are ~1 second

Let me know if you spot anything else! 🦅 👁️

gregrickaby committed:
update ignore paths

gregrickaby committed:
formatting

gregrickaby committed:
run format on staged php files

gregrickaby committed:
don't force purge during development

gregrickaby committed:
remove test scripts

gregrickaby committed:
phpcbf formatting

gregrickaby committed:
move all "fix" commands to format:*

gregrickaby committed:
use new format scripts

gregrickaby committed:
update scripts. closes #602

coreymcollins commented:
Thanks for these updates – I'll test these out today & tomorrow.

One thing of note: I'm not seeing the sprites on the frontend after running the build scripts. This is because `svg-spritemap-webpack-plugin` adds a prefix of `sprite-` to the SVG IDs, which we're not using.

We either need to set the prefix rule to `false` or update our SVG display functions. My preference would be for removing the prefix, not sure if you feel the same way?

gregrickaby committed:
update order. remove lint:md:js

gregrickaby committed:
don't prefix sprite IDs

gregrickaby commented:
@coreymcollins Nice catch! I've set `prefix: false` in `webpack.config.js`. 

Let me know if you find anything else! 🦅 👁️

coreymcollins commented:
I'm not sure how important this is to our workflow, but I'm noticing that SVG files aren't removed from the `/build/images/icons/` directory if you remove them from the corresponding `/src/` directory during `npm run watch` or `npm run start`.

Steps to reproduce:
1. Add an SVG to the `src` directory
2. Verify that it is copied to the `build` directory and added to the `sprite.svg` file
3. Remove an SVG from the `src` directory
4. Verify that the SVG isn't removed from the `build` directory but it IS removed from the `sprite.svg` file

This may not be a thing that's really a problem, because when you run `npm run build` (as we do on deployments) everything in the `build` directory is going to be dumped and rebuilt anyway, so you'll never see the phantom SVG.

But, if you're working in your local environment you won't see the SVG disappear from the `build` directory when you remove it from the `src` directory.

coreymcollins commented:
One more SVG thing!

I think we need to pull `svgo` into our `webpack.config.js` file, as the SVGs aren't stripped properly during `run` or `watch` and can't be altered with our SVG display functions currently.

Currently, the `build` versions of the SVGs output like:
```
<!-- Generated by IcoMoon.io -->
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" width="24" height="28" viewBox="0 0 24 28">
<title>pause</title>
<path fill="#fff" d="M24 3v22c0 0.547-0.453 1-1 1h-8c-0.547 0-1-0.453-1-1v-22c0-0.547 0.453-1 1-1h8c0.547 0 1 0.453 1 1zM10 3v22c0 0.547-0.453 1-1 1h-8c-0.547 0-1-0.453-1-1v-22c0-0.547 0.453-1 1-1h8c0.547 0 1 0.453 1 1z"></path>
</svg>
```

While we'd need them like this to be able to fully utilize them:
```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 28"><path d="M24 3v22c0 .547-.453 1-1 1h-8c-.547 0-1-.453-1-1V3c0-.547.453-1 1-1h8c.547 0 1 .453 1 1zM10 3v22c0 .547-.453 1-1 1H1c-.547 0-1-.453-1-1V3c0-.547.453-1 1-1h8c.547 0 1 .453 1 1z"/></svg>
```

To replicate:
1. Display an SVG in a page using:
```
_s_display_svg( array(
    'icon' => 'pause',
    'height' => '120',
    'width' => '120',
    'fill' => '#ff0000',
));
```
2. Note that the fill color isn't actually being recognized, so the SVG doesn't display.
3. Add this to `webpack.config.js`:
```
new ImageminPlugin( {
    svgo: {
        plugins: [
            { removeViewBox: false },
            { removeDimensions: true },
            {
                removeAttrs: {
                    attrs: '*:(stroke|fill):((?!^none$).)*',
                },
            },
        ],
    },
} ),
```
4. Run `npm run start` or `npm run watch`
5. Note the change to the SVG file itself
6. Note that the SVG display function now works and you can set the color of the SVG

coreymcollins committed:
Removes dark mode from Tailwind config

We aren’t using this out of the box and have a branch dedicated to this, plus keeping it adds about a second to our active build speeds

gregrickaby committed:
add SVGO to development webpack config

gregrickaby committed:
add clean-webpack-plugin. remove clean script and wp prettier.

gregrickaby committed:
clean build directory for each rebuild

gregrickaby committed:
add @wordpress/prettier-config

gregrickaby committed:
match with prettier config. allow tailwind rules.

gregrickaby committed:
formatting

gregrickaby commented:
@coreymcollins Thanks for all the feedback here. It's really appreciated. 🦅 👁️ 

1. The [edge-case SVG issue](https://github.com/WebDevStudios/wd_s/pull/618#issuecomment-812186210) has been resolved by replacing `npm run clean` with [CleanWebpackPlugin](https://www.npmjs.com/package/clean-webpack-plugin). Adding/removing SVG's from the `/src/images/icons` directory is mirrored in the `build` directory during active development.
2. I've added SVGO to the development Webpack config. Please verify this has fixed [your other issue](https://github.com/WebDevStudios/wd_s/pull/618#issuecomment-812186210).
3. In addition to those bug fixes, I've also extended the WordPress SCSS standards and massaged the `.stylelintrc.json` file rules to better work with Prettier. They were fighting each other over quotes and spacing. Finally, I added Tailwindcss rules to ignore list. See https://github.com/WebDevStudios/wd_s/pull/618/commits/83349abafe9b42e6ca585274e10040dbbd419c70

We have our Coding Standards meeting 11:00 ET today, looking forward to showing this off.

Let me know if you find anything else!

coreymcollins commented:
Lefthook doesn't appear to be working anymore.

I made some changes in a PHP file to break some spacing rules – on `main`, I get this when running `npx lefthook run pre-commit`:
![screenshot of Lefthook run on main branch](https://d.pr/i/XxUUaH+)

With the same change on this branch, I get this:
![screenshot of Lefthook run on this feature branch](https://d.pr/i/N5LpSs+)

If I remove this chunk from the bottom:
```
packagejson:
  glob: './package.json'
  run: npm run lint:pkg-json {staged_files}
```

Everything runs as expected – though, formatting is a bit off when the file is "fixed":
![lefthook results after formatting PHP](https://d.pr/i/ogRFZX+)

With our Sass partials, should Lefthook be formatting these if there are issues in the files?

If I write some Sass like this:
![Sass with some errors](https://d.pr/i/mTuN7T+)

The various spacing issues are not resolved/formatted. These also aren't being picked up when running `npm run lint:css`, though, so that may be the root of the problem.

coreymcollins commented:
Looking at source-maps this morning and they don't seem to be pointing where I'd expect them to.

For instance, on this heading:
![source-map screenshot](https://d.pr/i/h6erpE+)

I'm expecting it to point to the individual partial where these styles are, but it's pointing at the top-level `index.scss` partial where all of the imports happen.

gregrickaby committed:
help phpcs find composer

gregrickaby committed:
fix broken config. remove duplicate keys

gregrickaby committed:
use NPX and ./vendor for linting vs wp-scripts.

we need to pass {staged_files}, but both NPM scripts and Composer have paths specified

gregrickaby committed:
remove packagejson script

it runs even when package.json isn't staged. i'll revisit this later...

gregrickaby committed:
require empty lines

gregrickaby committed:
tweak format:css and format:js

run prettier first! the run with eslint --fix and stylint --fix

gregrickaby committed:
use stylelint-config-prettier. remove dup rule.

gregrickaby committed:
add npm run format:md script

gregrickaby commented:
@coreymcollins I believe I've resolved all the lefthook + linting issues. Please re-test and confirm.

I've spent the last hour looking into the source-map issue, and have not yet found a solution. It's def being caused by us using Sass partials. I'll continue investigating next week. 

FYI: the source-map experience you describe, is the same experience in the current `main` branch too.

gregrickaby committed:
add necessary loaders and plugins

gregrickaby committed:
process styles (with source-maps) and svgs

gregrickaby committed:
add file-loader for font support

gregrickaby committed:
use file-loader to deal with fonts

gregrickaby committed:
remove comment block

gregrickaby commented:
@coreymcollins After a deep dive this morning, I've figured it out how to get source-maps working correctly! Essentially, we need to enable source-maps for _all_ style loaders (CSS, PostCSS, and Sass):

![screenshot](https://dl.dropbox.com/s/cutxtqa0rxulvic/Screen%20Shot%202021-04-05%20at%2013.00.04.png?dl=0)

This doesn't come for free though. I've overridden half of the default @wordpress/scripts functionality 🙃, since the `webpack.config.js` file that ships with @wordpress/scripts only enables source-maps for CSS. **It's not complex, but does put the burden to maintain on WDS:**

```js
	module: {
		rules: [
			{
				test: /\.(sa|sc|c)ss$/,
				exclude: '/node_modules',
				use: [
					MiniCSSExtractPlugin.loader,
					{
						loader: 'css-loader',
						options: {
							sourceMap: true,
						},
					},
					{
						loader: 'postcss-loader',
						options: {
							sourceMap: true,
						},
					},
					{
						loader: 'sass-loader',
						options: {
							sourceMap: true,
						},
					},
				],
			},
			{
				test: /\.svg$/,
				use: [ '@svgr/webpack', 'url-loader' ],
			},
			{
				test: /\.(woff|woff2|eot|ttf|otf)$/,
				use: [
					{
						loader: 'file-loader',
						options: {
							name: 'fonts/[name].[hash:8].[ext]',
						},
					},
				],
			},
		],
	},
```

This solution also introduces 6 additional `devDependencies`. 😬  We need to decide as a group if we're comfortable maintaining such an opinionated `webpack.config.js` or figure out[ how to merge the config objects](https://github.com/survivejs/webpack-merge) (from WordPress and ours) so we can just add `sourceMap` support.

gregrickaby committed:
bump tailwind to v2.1.1

gregrickaby committed:
enable JIT mode

gregrickaby committed:
bump @wordpress prettier and scripts

aubreypwd commented on .phpcs.xml.dist:
```diff
@@ -3,6 +3,8 @@
 
 	<rule ref="WebDevStudios"/>
```
I wonder if you should use WordPress is doing here too? I'd also like to "reboot" PHP Coding Standards to match WordPress as much as possible and e.g. configure things like short-array-syntax, etc either at the config level (here) —especially if it's just this ONE config— or create a new package that has e.g. this tweak that you can install.

gregrickaby replied:
@aubreypwd Sounds good. I'll remove it and sprinkle in short-array-syntax.

aubreypwd reviewed (approved):
Damn Greg!

The only thing I see here is the use of our PHP Coding Standards, I wonder if maybe we should switch to WordPress' PHP Coding standards here too and just config things like `[]` e.g at the config level?

Curious what your thoughts are here! 

Great stuff 🙌 though!

aubreypwd reviewed (changes requested):
## Testing (Coding Standards)
- Mixing tabs/spaces in SASS files does not light up (was able to commit)
- In SASS `!important` didn't light up
- Using `underscore_style_functions` in JS didn't get flagged like they used t
- Proxying `https://wdunderscores.test/` on my machine didn't seem to work `https://localhost:3000/` would reload but didn't load anything for me saying `localhost didn’t send any data.` `http://localhost:3001/` shows me a UI though
- PHPCS lights up 🙌 
- JS lights up 🙌 
- Committing bad PHP didn't work 🙌 
- Wasn't able to commit bad JS 🙌 

```
❯ node --version && npm --version
v14.16.0
6.14.11
```

Note, `npm run format` did fix the SASS stuff like tabs/spaces, just didn't light up.

gregrickaby committed:
revert JIT support

this feature is out-of-scope for this PR. I was just playing around!

gregrickaby committed:
rename to phpcs.xml.dist

gregrickaby committed:
check in phpcompat.xml.dist

gregrickaby committed:
bring in setup from WordPress github

gregrickaby committed:
remove composer path

gregrickaby committed:
add PHP Sniffer & Beautifier to list

gregrickaby committed:
remove extra path

gregrickaby committed:
update standard filepath. add -s

gregrickaby committed:
add formatting commands

gregrickaby committed:
add declartion-no-important rule

gregrickaby committed:
add camelCase rule

gregrickaby commented:
@aubreypwd 

- Added formatting commands to Lefthook. The tabs vs. spaces issues should be taken care of by Prettier before a commit is pushed**
- Added `declaration-no-imporant` rule to `.stylelintrc.json`
- Added `camelCase` rule to `.eslintrc.js`

I'm not able to duplicate the proxy issue. My testing with `npm run watch` on `https://wdunderscores.test:3000/` works fine, and I'm seeing changes.

_** There is an issue with Tower and git hooks. Tower doesn't use the shell, so it's not aware of `PATH`, therefore cannot find Lefthook. See: https://github.com/Arkweid/lefthook/issues/99_

Workarounds are:

1. Use git via CLI 
2. Use git via your IDE
3. Install Lefthook globally
4. Create a [config file for Tower](https://www.git-tower.com/help/guides/integration/environment/mac)

gregrickaby committed:
add postcss-loader

gregrickaby committed:
add stylelint and postcss-reporter

gregrickaby committed:
add eslint-webpack-plugin

gregrickaby committed:
enable eslint-webpack-plugin

gregrickaby committed:
add stylelint-webpack-plugin

gregrickaby committed:
offload stylelint-ing to webpack

gregrickaby committed:
enable stylelint-webpack-plugin

gregrickaby commented:
I've added both `eslint-webpack-plugin` and `stylelint-webpack-plugin` to Webpack. This enables the real-time reporting of warnings and errors in the command line:

![screenshot](https://dl.dropbox.com/s/vf3y0tlzo6m0nti/Screen%20Shot%202021-04-13%20at%2009.09.44.png?dl=0)

Note: if there is an error, the site will not build. This [can be disabled](https://webpack.js.org/plugins/eslint-webpack-plugin/#failonwarning) though! Let's chat about it.

gregrickaby committed:
Merge branch 'main' into feature/coding-standards

coreymcollins commented:
How would the not building on error work in a production environment, when running through our Buddy process? I'm curious if "doesn't build" means you get a broken-looking site (i.e., some assets were rebuilt but others were not), or if it means nothing *new* builds and the old assets are still in place/not overwritten?

gregrickaby committed:
move php compat into phpcs.xml.dist

gregrickaby committed:
use phpcs.xml.dist for compat

gregrickaby committed:
dependency bump

gregrickaby committed:
rename to .phpcs.xml

gregrickaby committed:
point at updated file name

gregrickaby committed:
phpcs compatiblity formatting fixes

gregrickaby commented:
@coreymcollins I apologize, I should have been more clear. The build just exits until the error is fixed. With Buddy, it would stop and throw an error in a Slack Firehose channel.

- On a "fresh" build, JS/CSS assets will not be generated and the script just exits. (POT file runs, because that's a composer script)
- On a "rebuild", assets that were successfully generated prior are not overwritten/deleted.

![screenshot](https://dl.dropbox.com/s/84gb2h6eu0hozks/Screen%20Shot%202021-04-13%20at%2013.24.56.png?dl=0)

aubreypwd reviewed (commented):
- Proxying still doesn't work for me, I don't have any configuration that I don't think would conflict, I'm not sure why it doesn't work "on my machine."

## More and more rules...

> Added formatting commands to Lefthook. The tabs vs. spaces issues should be taken care of by Prettier before a commit is pushed**
> Added declaration-no-imporant rule to .stylelintrc.json
> Added camelCase rule to .eslintrc.js

It's great that we added these rules here, but this is where I was hoping "rebooting" coding standards to follow WP would take care of these (and others, if there are others). I find it odd that it's not taking care of these, I'm not sure I'm a fan of having to add all these caveats to config files... especially as they add up project to project (including wd_s here)...

...that also being said I did flag _just these few things_. I did not test _every_ pattern (_rule_) I could think of that are in line with what our coding standards _used_ to look like. Not sure what other rules have changed/don't work but it looks like they are adding up...

...that also being said I think this begs a couple questions:

1. Should we NOT tweak these things and simply allow what WP is doing guide us here, e.g. if it's cool with `!important` should we follow along?
2. If not (and we want to tweak them or it's a growing # of them e.g. not just one or two) I think this would warrant creating a package to add these tweaks vs using config files

## Commit Hook or PR Check step?

> There is an issue with Tower and git hooks. Tower doesn't use the shell, so it's not aware of PATH, therefore cannot find Lefthook

This has me worried. A lot of people use GUI clients...including me. I wonder if this means that maybe final formatting/linting needs to happen at the PR level (another option I've been mulling over since we decided to reboot CS): could we maybe do this linting/formatting at the PR level e.g. via something Buddy could run/etc vs. at the commit hook. E.g.:

- You commit code however you want, even code that's "bad", sure go ahead
- You have linting to guide you still, and if you like it, auto-formatting configs to help you out too on save, etc (but let's just say it's someone who doesn't like auto-formatting or linting)
- Once you open a PR it will run an check and this check will:
	- Try and auto-format your code for you on the files you changed (still not sure I'm a fan of this, but on new builds it may make more sense) and not bug you since we can fix it for you here
	- Run all the linting stuff afterwards (hopefully auto-formatting step caught a bunch of stuff for you)
	-  If the linting steps at this point still **do not** checkout, report back to the PR that it doesn't pass the check and (if we can do it) have the PR marked back to Draft (so developers would have to see this check pass to submit the PR to us)

The tower issue changed my perspective a bit on controlling things when the developer develops (e.g. commit hooks). PR creation (vs the commit hook) could try and auto-format for you and check (lint/etc) your code _there (in the PR check steps)_ vs while your developing. I'm an advocate of giving the developer more freedom on their computer and make it hard for developers to create a PR that doesn't follow our rules. Also I'm an advocate for people who like auto-formatting and want to use it while they are developing to help them along, but if you don't want to use it you don't have to—but then you have to fix your code before you can submit a PR (I'm one of those people who are willing to take on that task).  It would use the same tools that are installed in the project that would report to your editor in linting and the same auto-formatting config you might also have on in your editor. 

If all the checks and auto-formatting attempts (same thing we're trying to do on commit hooks) happen at a PR check step then we rid ourselves of having to make our configurations super-compatible with editor(s): Vim, Sublime, PHPStorm, etc vs. VSCode. Of course we would have tools in the project to help your editors out, but we don't have to worry about the person's computer being perfectly configured to get a PR created properly (e.g. a perfectly formatted PR). We can just (IMO) do that in a PR check step...

1. Try to auto-format it for you
2. Lint the final result

(All stuff we would do in commit hooks)

So if you have auto-formatting on and linting on in your editor, you're PR creation step should go super smooth. If you don't then you might have things to fix (but I anticipate most people will at least have linting on, but who knows).

**What do you think?** I know this is drastically different than doing this on commit hooks, but it has the same result: **No submitted PR's that don't have formatting issues.**

This gives the developer more freedom on their computer. E.g. I can have auto-formatting off  (because I don't like it personally—but have linting on) and be able to work without having to figure out how this project wants my code to look. Or I can be the type of person who wants auto-formatting to just do it for me in my editor and I will adapt. When I create a PR it will auto-format my code and lint my code. If it checks out a PR is created, if not it's punted back to a Draft.

Someone like @itsamoreh could have linting and auto-formatting on in his editor, and write code according to the formatting we've configured in the project as he goes. He can get the benefits of auto-formatting and it's highly likely when he gets to the PR process he will have more success. Because he had auto-formatting on, his code will likely pass the steps the PR hook will do. 

...if my idea here is even possible. 

### Caveats of my idea here:

- When the auto-formatter runs in the assertion step, it would have to be able to commit any changes for you and push it up to the branch you created (which I think could be possible)
- When this commit goes up to GH you would have to know to always update your branch e.g. a `git pull` to pull that auto-format attempt

And to be clear, I am suggesting we auto-format code in an assertion NOT in your editor (unless you want it to via auto-formatting and linting) and NOT in a commit hook. You can commit code just as you would, but to open a PR a check would have to pass where we would attempt to auto-format your code for you, and only if it fails then you would have to fix it yourself in your editor where things like `npm run watch` and `npm run lint*` tasks could be used to look at where you need to make corrections. If you have auto-formatting on in your editor (or at the very least follow linting) though you shouldn't have a problems getting a PR to pass...

richaber commented on .phpcs.xml:
```diff
@@ -0,0 +1,46 @@
+
+	<!-- Only sniff the theme. -->
+	<file>./</file>
+
```
I think we should include a few more "helpful" rules, like allowing file name exceptions for themes, defining the text domain and function prefixes. 

The customizable WPCS sniffs are all listed in the [WPCS wiki](https://github.com/WordPress/WordPress-Coding-Standards/wiki/Customizable-sniff-properties).

```
<!-- Allow for theme specific exceptions to the file name rules based
	 on the theme hierarchy. -->
<rule ref="WordPress.Files.FileName">
	<properties>
		<property name="is_theme" value="true"/>
	</properties>
</rule>

<!-- Verify that everything in the global namespace is prefixed with a theme specific prefix.
	 Multiple valid prefixes can be provided as a comma-delimited list. -->
<rule ref="WordPress.NamingConventions.PrefixAllGlobals">
	<properties>
		<property name="prefixes" type="array" value="_s" />
	</properties>
</rule>

<!-- Verify that the text_domain is set to the desired text-domain.
	 Multiple valid text domains can be provided as a comma-delimited list. -->
<rule ref="WordPress.WP.I18n">
	<properties>
		<property name="text_domain" type="array" value="_s"/>
	</properties>
</rule>

```

richaber replied:
Maybe set the minimum supported WP version too.
Looks like our style.css says WP 5.5
```
	<!-- Set the minimum supported WP version. This is used by several sniffs. -->
	<config name="minimum_supported_wp_version" value="5.5"/>
```

gregrickaby replied:
Thanks, @richaber 

During my first pass at testing I get:

```text
The "_s" prefix is too short. Short prefixes are not unique enough and may cause name collisions with other code.
```

Due to this rule:

```xml
<!-- Verify that everything in the global namespace is prefixed with a theme specific prefix.
	 Multiple valid prefixes can be provided as a comma-delimited list. -->
<rule ref="WordPress.NamingConventions.PrefixAllGlobals">
	<properties>
		<property name="prefixes" type="array" value="_s" />
	</properties>
</rule>
```

Thoughts here?

richaber replied:
```
<rule ref="WordPress">
	<exclude name="WordPress.NamingConventions.PrefixAllGlobals.ShortPrefixPassed"/>
</rule>
```

gregrickaby replied:
@richaber Got it! See https://github.com/WebDevStudios/wd_s/pull/618/commits/8c1b9861e3e7ef4c9d444846160960275f935669

Thank you, sir!

gregrickaby committed:
remove folders that don't exist

gregrickaby committed:
specify node version. closes #631

gregrickaby committed:
add wp-cli to prereq. add toc.

gregrickaby committed:
bump theme version

gregrickaby committed:
make phpcs more theme specific

gregrickaby committed:
add wp cli to list. closes #629

gregrickaby committed:
add prefix name rule (and allow short prefixes)

gregrickaby committed:
specify wordpress's prettier. closes #630

gregrickaby committed:
add composer/installers

jrfoell reviewed (approved):
Thanks for you work on this @gregrickaby

aubreypwd reviewed (approved):
Per our demo and chat today this is good to go! Great work Greg!

salcode commented:
These notes are from commit 5b2e695

## composer.lock

During `npm i`, I got the message

> Warning: The lock file is not up to date with the latest changes in composer.json. You may be getting outdated dependencies. It is recommended that you run `composer update` or `composer update <package name>`.

indicating `composer.lock` is not up to date with `composer.json` 

Running `composer update` did not introduce any changes (other than updating the `content-hash` in `composer.lock`), we can update this hash (and just this hash) with 

```
composer update --lock
```

which will eliminate the warning message.

## In Editor Linting Errors

In Vim, running the [ALE plugin](https://github.com/dense-analysis/ale) with the settings I already had in place, I'm seeing linting errors flagged but changes are not automatically applied.

e.g. 
![image](https://user-images.githubusercontent.com/5194588/115081970-21f62680-9ed3-11eb-979d-b158ad5596a3.png)

### Exception to Flagging Errors

In `src/scss/base/_headings.scss` when I changed the indentation on a line from tabs to spaces, it was not flagged.

![image](https://user-images.githubusercontent.com/5194588/115082139-6681c200-9ed3-11eb-82a6-c05137b1d872.png)

Other errors in the same file _were_ flagged.

![image](https://user-images.githubusercontent.com/5194588/115082212-8618ea80-9ed3-11eb-9456-29441b80eb4d.png)

## Linting on Commit

When I ran `git commit` the linting errors that could be automatically corrected were and my commit was rejected 👍

It was a little confusing that I then had both staged (my original changes) and unstaged (the linting fixes) changes. I think this _is_ the best way to handle things but I can imagine this may be a stumbling point for those new to automated linting. We may want to call this out specifically when training team members to use this.

### Unflagged Linting Error was Still Fixed

The SCSS linting error I introduced by changing an indent from tab to spaces that was not flagged in the editor (as mentioned earlier in this comment), _was_ fixed during the correcting on commit.

This may be an issue in my editor failing to flag the space/tab problem 🤷‍♂️ .

## Overall Impression

Overall, I think this is an impressive setup that should help team members avoid wasting time correcting linting errors. 👍

Great work!

gregrickaby commented:
@salcode Thanks for the feedback and overall sign-off. 🥳 

### Tabs vs Spaces

This isn't a _rule_, but rather an _opinion_ listed in both `.editorconfig` and  `.prettierrc.js`. Therefore, they will never be "flagged" by any IDE as an error/warning. They will be re-formatted though-- when Prettier runs.

### Vim

Unfortunately, I'm not knowledgeable enough about VIM or the plugin you mentioned to provide any assistance. I'm glad that the `pre-commit` hooks caught the issues and took care of them.  

### Demo

I'll be doing a demo on Tuesday's FEE/BEE scrum, and would happy to walk through the commit hook process. See ya then! 👍🏻

gregrickaby committed:
add wp-cli as dependency. configure pot script.

gregrickaby committed:
dependency bump

gregrickaby committed:
remove platform check

gregrickaby committed:
install with composer v2 (it works!)

gregrickaby committed:
remove wp cli dependency

gregrickaby commented:
@coreymcollins I updated the Assertions in Buddy and ran them against this PR. Green light! 🟢 

I humbly await your sign-off so we can get this merged in. 

Thanks!

gregrickaby commented:
Just noting that Corey and I spoke in Slack and he's 👍🏻
